README
NAME
    Crypt::Polybius - implementation of the Polybius square

SYNOPSIS
       use Crypt::Polybius;
   
       #      1    2    3    4    5
       # 1    A    B    C    D    E
       # 2    F    G    H    I/J  K
       # 3    L    M    N    O    P
       # 4    Q    R    S    T    U
       # 5    V    W    X    Y    Z
       #
       # ATTACK  ->  11 44 44 11 13 25
       # AT      ->  11 44
       # DAWN    ->  14 11 52 33
   
       my $square = Crypt::Polybius->new;
   
       print $square->encipher('Attack at dawn.'), "\n";

DESCRIPTION
    This module provides an implementation of the Polybius square, or Polybius
    checkerboard. This cipher is not cryptographically strong, nor completely
    round-trip-safe.

  Constructor
    `new(%attributes)`
        Moose-like constructor.

  Attributes
    `square`
        An array of arrays of letters. The default is:

           [
              [qw/ A B C D E /],
              [qw/ F G H I K /],
              [qw/ L M N O P /],
              [qw/ Q R S T U /],
              [qw/ V W X Y Z /],
           ]

    `substitutions`
        A hashref of strings of pre-encipher substitutions to make. The
        default is:

           { "J" => "I" }

        You should only use upper-case letters.

  Methods
    `encipher($str)`
        Enciphers a string and returns the ciphertext.

    `decipher($str)`
        Deciphers a string and returns the plaintext.

    `preprocess($str)`
        Perform pre-encipher processing on a string. `encipher` calls this, so
        you are unlikely to need to call it yourself.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Crypt-Polybius>.

SEE ALSO
    <http://en.wikipedia.org/wiki/Polybius_square>.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Crypt-Polybius
==============

Created:      2014-05-23
Home page:    <https://metacpan.org/release/Crypt-Polybius>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Crypt-Polybius>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.002	2014-05-25

 - Added: Support for the Greek alphabet.
 - Added: Support for the scrambling the alphabet with a password.
 - Much refactoring, splitting things out into roles, making stuff more
   immutable, and adding type constraint to attributes.

0.001	2014-05-23	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "implementation of the Polybius square",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Crypt-Polybius",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => {
                                       requires => {
                                         "Const::Fast" => 0,
                                         "Moo" => "1.002000",
                                         "Moo::Role" => 0,
                                         "namespace::sweep" => 0,
                                         "perl" => 5.008,
                                         "Text::Unidecode" => 0,
                                         "Types::Common::Numeric" => 0,
                                         "Types::Common::String" => 0,
                                         "Types::Standard" => 0,
                                       },
                                     },
                        test      => { requires => { "Test::Modern" => 0.007 } },
                      },
  "provides"       => {
                        "Crypt::Polybius"                 => { file => "lib/Crypt/Polybius.pm", version => 0.002 },
                        "Crypt::Polybius::Greek"          => { file => "lib/Crypt/Polybius/Greek.pm", version => 0.002 },
                        "Crypt::Role::CheckerboardCipher" => { file => "lib/Crypt/Role/CheckerboardCipher.pm", version => 0.002 },
                        "Crypt::Role::GreekAlphabet"      => { file => "lib/Crypt/Role/GreekAlphabet.pm", version => 0.002 },
                        "Crypt::Role::LatinAlphabet"      => { file => "lib/Crypt/Role/LatinAlphabet.pm", version => 0.002 },
                        "Crypt::Role::ScrambledAlphabet"  => { file => "lib/Crypt/Role/ScrambledAlphabet.pm", version => 0.002 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=Crypt-Polybius",
                                        },
                        homepage     => "https://metacpan.org/release/Crypt-Polybius",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-crypt-polybius.git",
                                          web  => "https://github.com/tobyink/p5-crypt-polybius",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Crypt-Polybius/project",
                      },
  "version"        => 0.002,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

